py: Add support for function doc strings.

dpgeorge opened the pull request:
This PR adds support for function doc strings.  It's disabled by default.

```py
def f():
    "doc string"
    pass

print(f.__doc__)
```

TODO:
- [x] add tests
- [x] ~~maybe have a separate compile-time option for this~~ no, let's keep it under the existing MICROPY_ENABLE_DOC_STRING option
- [ ] check that the native .mpy ABI is unchanged

dpgeorge committed:
py: Add support for function docstrings.

This stores the docstring in the function object at compile time.

Signed-off-by: Damien George <damien@micropython.org>

jimmo commented on py/compile.c:
```diff
@@ -782,6 +782,32 @@ STATIC qstr compile_funcdef_helper(compiler_t *comp, mp_parse_node_struct_t *pns
+    #if MICROPY_ENABLE_DOC_STRING
+    // look for the first statement
+    mp_parse_node_t pn = pns->nodes[3]; // the function body/suite
+    if (MP_PARSE_NODE_IS_STRUCT_KIND(pn, PN_suite_block_stmts)) {
```
This check already seems to be done by `check_for_doc_string`?

i.e. this whole thing could be a call to `check_for_doc_string`, the only difference is how it emits the store of `__doc__` ?

dpgeorge replied:
Good idea.  I've now done this code refactor.

jimmo commented on py/compile.c:
```diff
@@ -782,6 +782,32 @@ STATIC qstr compile_funcdef_helper(compiler_t *comp, mp_parse_node_struct_t *pns
     // compile the function definition
     compile_funcdef_lambdef(comp, fscope, pns->nodes[1], PN_typedargslist);
 
+    #if MICROPY_ENABLE_DOC_STRING
```
Can we update the comment in `mpconfig.h` to explain that this enables docstrings on functions, modules, and classes?

(I think we should keep them as one combined option)

dpgeorge replied:
Comment updated.

dpgeorge commented:
I really want to find a way to make this use less RAM for functions that don't have doc strings.  Because at the moment all functions increase by 1 GC block size, regardless of whether they use the new doc string slot.

But I don't know how to do that in a simple way.  It probably requires a separate function type like `mp_type_fun_bc_with_doc_string` and the ability to upgrade a normal function to this type when the doc string entry is stored to (and eg repurpose the `const byte *bytecode` entry in `mp_obj_fun_bc_t` to point to a separate struct that points to the doc string and the bytecode).

If we can find a way to do that then we could enable this feature on the large ports, because it would only cost something if docs strings are used.

Also could extend things to allow any attribute to be stored to a function object, not just the `__doc__` entry.  Doc strings would then just be one user of this feature (to store attrs to functions).

Josverl commented:
Out of interest wrt to the Micropython-stubs project I'd like to understand what the goal is for this feature

Is it targeting stdlib/micropython code, frozen (. Mpy) code, or regular micropython.py modules?

I assume the interface would be the cpython style `__doc__` property, is that correct?

Is there something similar considered for modules or classes.?

dpgeorge committed:
py/mpconfig: Enable docstrings at the everything config level.

This will be enabled on the unix coverage build.

Also add a test for this feature.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
py/compile: Factor docstring code.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
py/mpconfig: Update MICROPY_ENABLE_DOC_STRING docs.

Signed-off-by: Damien George <damien@micropython.org>
